subsys/esb/esb_dppi.c

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <hal/nrf_egu.h>
+#include <hal/nrf_radio.h>
+#include <hal/nrf_timer.h>
+
+#include <nrfx_dppi.h>
+
+#include <zephyr/logging/log.h>
+
+#include "esb_peripherals.h"
+#include "esb_ppi_api.h"
+
+LOG_MODULE_DECLARE(esb, CONFIG_ESB_LOG_LEVEL);
+
+static uint8_t radio_address_timer_stop;
+static uint8_t timer_compare0_radio_disable;
+static uint8_t timer_compare1_radio_txen;
+static uint8_t disabled_egu;
+static uint8_t egu_timer_start;
+static uint8_t egu_ramp_up;
+
+static nrf_dppi_channel_group_t ramp_up_dppi_group;
+
+void esb_ppi_for_txrx_set(bool rx, bool timer_start)
+{
+	uint32_t channels_mask;
+
+	nrf_egu_event_clear(ESB_EGU, ESB_EGU_EVENT);
+	nrf_egu_event_clear(ESB_EGU, ESB_EGU_DPPI_EVENT);
+
+	nrf_egu_publish_set(ESB_EGU, ESB_EGU_EVENT, egu_timer_start);
+	nrf_egu_publish_set(ESB_EGU, ESB_EGU_DPPI_EVENT, egu_ramp_up);
+
+	nrf_dppi_channels_include_in_group(NRF_DPPIC, BIT(egu_ramp_up), ramp_up_dppi_group);
+
+	nrf_egu_subscribe_set(ESB_EGU, ESB_EGU_DPPI_TASK, egu_timer_start);
+	nrf_radio_subscribe_set(NRF_RADIO, rx ? NRF_RADIO_TASK_RXEN : NRF_RADIO_TASK_TXEN,
+				egu_ramp_up);
+	nrf_dppi_subscribe_set(NRF_DPPIC,
+				nrf_dppi_group_disable_task_get((uint8_t)ramp_up_dppi_group),
+				egu_ramp_up);
+	nrf_egu_subscribe_set(ESB_EGU, ESB_EGU_TASK, disabled_egu);
+
+	if (timer_start) {
+		nrf_timer_subscribe_set(ESB_NRF_TIMER_INSTANCE, NRF_TIMER_TASK_START,
+					egu_timer_start);
+	}
+
+	channels_mask = (BIT(egu_timer_start) |
+			 BIT(egu_ramp_up));
+
+	nrf_dppi_channels_enable(NRF_DPPIC, channels_mask);
+}
+
+void esb_ppi_for_txrx_clear(bool rx, bool timer_start)
+{
+	uint32_t channels_mask;
+
+	channels_mask = (BIT(egu_timer_start) |
+			 BIT(egu_ramp_up));
+
+	nrf_dppi_channels_disable(NRF_DPPIC, channels_mask);
+
+	nrf_egu_publish_clear(ESB_EGU, ESB_EGU_EVENT);
+	nrf_egu_publish_clear(ESB_EGU, ESB_EGU_DPPI_EVENT);
+
+	nrf_egu_subscribe_clear(ESB_EGU, ESB_EGU_DPPI_TASK);
+	nrf_radio_subscribe_clear(NRF_RADIO, rx ? NRF_RADIO_TASK_RXEN : NRF_RADIO_TASK_TXEN);
+	nrf_dppi_subscribe_clear(NRF_DPPIC,
+				 nrf_dppi_group_disable_task_get((uint8_t)ramp_up_dppi_group));
+	nrf_egu_subscribe_clear(ESB_EGU, ESB_EGU_TASK);
+
+	nrf_dppi_channels_remove_from_group(NRF_DPPIC, BIT(egu_ramp_up), ramp_up_dppi_group);
+
+
+	if (timer_start) {
+		nrf_timer_subscribe_clear(ESB_NRF_TIMER_INSTANCE, NRF_TIMER_TASK_START);
+	}
+}
+
+void esb_ppi_for_fem_set(void)
+{
+	nrf_egu_publish_set(ESB_EGU, ESB_EGU_EVENT, egu_timer_start);
+	nrf_timer_subscribe_set(ESB_NRF_TIMER_INSTANCE, NRF_TIMER_TASK_START,
+				egu_timer_start);
+
+	nrf_dppi_channels_enable(NRF_DPPIC, BIT(egu_timer_start));
```

Comment:
Aren't we doing the same here and in `esb_ppi_for_txrx_set` (subscribing timer start event to a channel to which ESB_EGU_EVENT publishes)? Couldn't the code be reused?

Reply:
I think there won't be a huge benefit in moving this part to separate function. In my opinion it will complicate readability of this code. I would like to keep it in this way

Reply:
I think we meant something different.The thing is I am not sure if there is any case, in which `esb_ppi_for_fem_set` is needed - it seems, that the same configuration is always done before by `esb_ppi_for_txrx_set` which would make this function redundant.

Reply:
However this is a minor thing, approving anyway